### PR TITLE
Allow specification of initial IPAM ring via --ipam-seed

### LINKED ISSUE #1998: Ability to specify initial IPAM ring seed
If I could do this
```
host1$ weave launch --name 00::01 --ipam-seed 00::01,00::02,00::03 $HOST1 $HOST3 $HOST3
host2$ weave launch --name 00::02 --ipam-seed 00::01,00::02,00::03 $HOST1 $HOST3 $HOST3
host3$ weave launch --name 00::03 --ipam-seed 00::01,00::02,00::03 $HOST1 $HOST3 $HOST3
```
it would be possible for me to spin up a cluster of parameterised VM images containing 'pre-concensed' weave nodes that could operate under partition from the outset.

### PR DESCRIPTION
**Pull request by awh**
Fixes #1998.

### CODE REVIEW

File: prog/weaver/main.go [hunk 1]
```diff
@@ -106,6 +107,7 @@ func main() {
 	mflag.StringVar(&dbPrefix, []string{"-db-prefix"}, "weave", "pathname/prefix of filename to store data")
 
 	mflag.StringVar(&trustedSubnetStr, []string{"-trusted-subnets"}, "", "Comma-separated list of trusted subnets in CIDR notation")
+	mflag.StringVar(&ipamSeedStr, []string{"-ipam-seed"}, "", "Comma-separated list of peer names used to form initial ring")
```
Comment (rade) [inline prog/weaver/main.go]: nothing anywhere in our docs mentions "ring". and rightly so. We need to explain this in terms that don't delve into implementation detail.

**Comment by awh** [comment]
Docs deliberately missing from this PR to avoid clashing with the documentation restructuring - will add some after that has merged.

**Comment by rade** [comment]
> Docs deliberately missing from this PR to avoid clashing with the documentation restructuring - will add some after that has merged.

Please don't defer the doc changes.

**Comment by awh** [comment]
Currently working on test breakage and some docs.

**Comment by awh** [comment]
Rebased, fixed tests, added docs.

**Comment by bboreham** [comment]
I found it striking that we make users go to all the trouble of formatting a peer-name that looks like a MAC.  We could just let them supply a number and interpret that as a peername internally.

File: ipam/allocator.go [hunk 2]
```diff
@@ -804,31 +806,50 @@ func (alloc *Allocator) persistRing() {
...
 	}
-	for _, cidrs := range alloc.owned {
-		for _, cidr := range cidrs {
-			alloc.space.Claim(cidr.Addr)
+
+	if nameFound {
+		if checkPeerName == alloc.ourName {
+			if ringFound {
+				if len(alloc.seed) != 0 {
+					alloc.infof("Found persisted ring data, ignoring supplied IPAM seed")
```
Comment (bboreham) [inline ipam/allocator.go]: maybe "persisted IPAM data", since we don't document "ring"

**Comment by awh** [comment]
> I found it striking that we make users go to all the trouble of formatting a peer-name that looks like a MAC. We could just let them supply a number and interpret that as a peername internally.

Totally agree - I was thinking of following up with another PR that lets us parse IPv6 style zero-runs e.g. ::1, ::2, ::3 - thoughts? It just means adjusting the peer name parses, should be pretty easy - I might have a stab at adding it quickly to this PR if I have time today.

**Comment by rade** [comment]
I don't get it. The peer name does show up in various places in the ui, in mac notation. There is no point having different ways of inputting it if the output remains the same.

**Comment by awh** [comment]
> There is no point having different ways of inputting it if the output remains the same.

Disagree - what's wrong with having a shorthand notation for use on the command line for specifying the values?

**Comment by rade** [comment]
> what's wrong with having a shorthand notation for use on the command line for specifying the values?

Because then you cannot correlate back the values in the UI.

btw, you could just hash any user supplied name into a MAC, which nicely illustrates the problem with this approach.

**Comment by awh** [comment]
> Because then you cannot correlate back the values in the UI.

You can, because there is an extremely simple rule that tells you that ::01 is equivalent to 00:00:00:00:00:01.

> btw, you could just hash any user supplied name into a MAC, which nicely illustrates the problem with this approach.

It doesn't, because humans can trivially reverse the mapping I'm talking about, unlike the hash you're suggesting.

**Comment by awh** [comment]
Here's the kind of scheme I mean (taken from https://en.wikipedia.org/wiki/IPv6_address#Recommended_representation_as_text):

* Leading zeros in each 16-bit field are suppressed. For example, 2001:0db8::0001 is rendered as 2001:db8::1, though any all-zero field that is explicitly presented is rendered as 0.
* "::" is not used to shorten just a single 0 field. For example, 2001:db8:0:0:0:0:2:1 is shortened to 2001:db8::2:1, but 2001:db8:0000:1:1:1:1:1 is rendered as 2001:db8:0:1:1:1:1:1.
* Representations are shortened as much as possible. The longest sequence of consecutive all-zero fields is replaced by double-colon. If there are multiple longest runs of all-zero fields, then it is the leftmost that is compressed. E.g., 2001:db8:0:0:1:0:0:1 is rendered as 2001:db8::1:0:0:1 rather than as 2001:db8:0:0:1::1.

File: ipam/allocator.go [hunk 3]
```diff
@@ -66,9 +67,10 @@ type Allocator struct {
 }
 
 // NewAllocator creates and initialises a new Allocator
-func NewAllocator(ourName mesh.PeerName, ourUID mesh.PeerUID, ourNickname string, universe address.Range, quorum uint, db db.DB, isKnownPeer func(name mesh.PeerName) bool) *Allocator {
+func NewAllocator(ourName mesh.PeerName, ourUID mesh.PeerUID, ourNickname string, seed []mesh.PeerName, universe address.Range, quorum uint, db db.DB, isKnownPeer func(name mesh.PeerName) bool) *Allocator {
```
Comment (rade) [inline ipam/allocator.go]: I think it's time for a Config struct.

**Comment by rade** [comment]
> Here's the kind of scheme I mean

Yes, please!

File: ipam/allocator.go (hunk 3 above)
Comment (awh) [inline ipam/allocator.go]: 2eeb754454a64026c85280ff2e55af37d04c8e0d

**Comment by awh** [comment]
> Yes, please!

weaveworks/mesh#23

**Comment by awh** [comment]
Will push a final update once the mesh PR is merged.

**Comment by awh** [comment]
Ready for final review.

File: site/ipam/overview-init-ipam.md [hunk 4]
```diff
@@ -22,6 +22,31 @@ The following automatic IP address managment topics are discussed:
...
+outset even under conditions of partition - no consensus is required.
+
+Alternatively, you can let weave determine the seed automatically via
+a consensus algorithm; since you don't need to provide it with a list
+of peer names anymore you can also let weave generate those randomly
+for you too. However, in order for weave to form a single consensus
+reliably you must now instead tell each peer how many peers there are
+in total either by listing them as target peers or via the
+`--init-peer-count` parameter.
+
```
Comment (rade) [inline site/ipam/overview-init-ipam.md]: It is not obvious that the paragraphs that follow only apply to the "configuration via consensus" options. Perhaps we should create two sub sections.

**Comment by rade** [comment]
I *think* the `--observer` flag works irrespective of whether the other peers use seeding or consensing.. That isn't obvious at all. That section currently (post #2060) mentions 'consensus', which would need replacing with more generic terminology, perhaps around "initialising IP address allocation" (btw, don't mention 'IPAM' inside the IPAM docs).

We may need to change the option name in the process. `--ipam-observer`, for starters.

`weave consense` is (now even more obviously) misnamed, since (I think) it can be used to initialise ipam on a peer from others that were seeded, not consensed. Or even when the peer itself has been given a `--ipam-seed` but perhaps not containing itself.

Finally, what happens if you specify more than one of `--ipam-seed`, `--init-peer-count`, and `--observer`?

File: db/boltdb.go [hunk 5]
```diff
@@ -50,11 +50,13 @@ func NewBoltDB(dbPathname string) (*BoltDB, error) {
...
 
-func (d *BoltDB) Load(ident string, data interface{}) error {
-	return d.db.View(func(tx *bolt.Tx) error {
+func (d *BoltDB) Load(ident string, data interface{}) (bool, error) {
+	found := true
+	return found, d.db.View(func(tx *bolt.Tx) error {
 		b := tx.Bucket(topBucket)
 		v := b.Get([]byte(ident))
 		if v == nil {
+			found = false
```
Comment (rade) [inline db/boltdb.go]: you are making a rather big assumption about evaluation here, namely that the `found` in `return found, ...` is evaluated *after* the `View` function has executed. The go spec may well guarantee that, but I'd rather avoid such subtlety.

**Comment by rade** [comment]
The IPAM status terminology - both as documented and as displayed (which is different! #2061) - no longer makes sense, since it assumes the status is mostly driven by consensus, which will no longer be the case post this PR.

**Comment by awh** [comment]
> We may need to change the option name in the process. --ipam-observer, for starters.

https://github.com/weaveworks/weave/pull/1990#issuecomment-191164562

**Comment by awh** [comment]
> you are making a rather big assumption about evaluation here

Negatory, I tested it :)

> The go spec may well guarantee that, but I'd rather avoid such subtlety.

Fair enough!

**Comment by rade** [comment]
> https://github.com/weaveworks/weave/pull/1990#issuecomment-191164562

ha ha. I was wondering how long it would take you to quote that ;)

**Comment by awh** [comment]
> I think the --observer flag works irrespective of whether the other peers use seeding or consensing..

Technically yes - from the point of view of an observer, explicitly seeded peers are indistinguishable from those that were seeded by consensus. This would imply mixed usage of seeding and consensus in the same cluster though, which I wouldn't recommend.

> weave consense is (now even more obviously) misnamed  since (I think) it can be used to initialise ipam on a peer from others that were seeded, not consensed.

See above - you shouldn't (IMO) be using `weave consense` in a seeded cluster.

> Or even when the peer itself has been given a --ipam-seed but perhaps not containing itself.

Even here, `weave consense` doesn't make sense - such a peer is already initialised. Note that `weave consense` does not guarantee that you own a piece of the ring, just that you have a copy of the ring. If it _did_ make such a guarantee it would be useful in this scenario, as you could run it to ensure that a freshly added peer could subsequently make headway under partition.

> what happens if you specify more than one of --ipam-seed, --init-peer-count, and --observer?

Currently `--observer` and `--init-peer-count` are mutually exclusive - I should add `--ipam-seed` to that set.

**Comment by awh** [comment]
> The IPAM status terminology - both as documented and as displayed (which is different! #2061) - no longer makes sense, since it assumes the status is mostly driven by consensus, which will no longer be the case post this PR.

Actually three of the five possible states remain relevant; peers launched with `--ipam-seed` will start in `ready` instead of `idle`. Seems fine to me - not sure how it could be changed.

**Comment by awh** [comment]
> The IPAM status terminology - both as documented and as displayed (which is different! #2061) - no longer makes sense, since it assumes the status is mostly driven by consensus, which will no longer be the case post this PR.

I have raised #2063 to cover.

**Comment by awh** [comment]
> ... Perhaps we should create two sub sections ... Perhaps refer to it as "one-off consensus" in most places.

Docs updated.

> Currently --observer and --init-peer-count are mutually exclusive - I should add --ipam-seed to that set.

Done.

**Comment by awh** [comment]
Remaining concerns to be addressed by #2065.

File: prog/weaver/main.go [hunk 6]
```diff
@@ -96,6 +96,7 @@ func main() {
 		datapathName       string
 		trustedSubnetStr   string
 		dbPrefix           string
+		ipamSeedStr        string
```
Comment (rade) [inline prog/weaver/main.go]: This should go higher up, near the other ipam related args.

File: prog/weaver/main.go [hunk 7]
```diff
@@ -134,6 +135,7 @@ func main() {
 	mflag.StringVar(&dbPrefix, []string{"-db-prefix"}, "weave", "pathname/prefix of filename to store data")
 
 	mflag.StringVar(&trustedSubnetStr, []string{"-trusted-subnets"}, "", "Comma-separated list of trusted subnets in CIDR notation")
+	mflag.StringVar(&ipamSeedStr, []string{"-ipam-seed"}, "", "Comma-separated list of peer names amongst which address space is shared initially")
```
Comment (rade) [inline prog/weaver/main.go]: ditto

File: test/127_ipam_seed_2_test.sh [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+assert_raises "timeout 10 cat <( start_container $HOST2 --name c2)"
+
+# Connect routers
+weave_on $HOST2 connect $HOST1
+
+# Check connectivity
+assert_raises "exec_on $HOST1 c1 $PING c2"
+assert_raises "exec_on $HOST2 c2 $PING c1"
+
+end_suite
```
Comment (rade) [inline test/127_ipam_seed_2_test.sh]: This file needs moving into the 3xx range.

**Comment by rade** [comment]
arg needs adding to weave usage. So does `--observer`.

File: prog/weaver/main.go [hunk 9]
```diff
@@ -467,6 +471,23 @@ func parseTrustedSubnets(trustedSubnetStr string) []*net.IPNet {
...
+func parseIPAMSeed(ipamSeed string) []mesh.PeerName {
+	peerNames := []mesh.PeerName{}
+	if ipamSeed == "" {
+		return peerNames
+	}
+
+	for _, peerNameStr := range strings.Split(ipamSeed, ",") {
+		peerName, err := mesh.PeerNameFromUserInput(peerNameStr)
+		if err != nil {
+			Log.Fatal("Unable to parse IPAM seed: ", err)
```
Comment (rade) [inline prog/weaver/main.go]: Does this show the piece of the input it was tripping over? It should.

**Comment by awh** [comment]
@rade all comments addressed, except for the one regarding add `--observer` to the weave script usage - I will push that directly to master.

**Pull request merged**